dhis-2/dhis-services/dhis-service-administration/src/test/java/org/hisp/dhis/maintenance/HardDeleteAuditTest.java

```diff
@@ -0,0 +1,89 @@
+package org.hisp.dhis.maintenance;
+
+import com.google.common.collect.Sets;
+import org.hisp.dhis.IntegrationTestBase;
+import org.hisp.dhis.audit.Audit;
+import org.hisp.dhis.audit.AuditQuery;
+import org.hisp.dhis.audit.AuditService;
+import org.hisp.dhis.audit.AuditType;
+import org.hisp.dhis.common.IdentifiableObjectManager;
+import org.hisp.dhis.maintenance.jdbc.JdbcMaintenanceStore;
+import org.hisp.dhis.organisationunit.OrganisationUnit;
+import org.hisp.dhis.trackedentity.TrackedEntityAttribute;
+import org.hisp.dhis.trackedentity.TrackedEntityInstance;
+import org.hisp.dhis.trackedentity.TrackedEntityInstanceService;
+import org.junit.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.test.context.ActiveProfiles;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import static org.awaitility.Awaitility.await;
+import static org.junit.Assert.assertEquals;
+
+@ActiveProfiles( profiles = { "test-audit" } )
+public class HardDeleteAuditTest
+    extends IntegrationTestBase
+{
+    private static final int TIMEOUT = 5;
+    @Autowired
+    private AuditService auditService;
+
+    @Autowired
+    private TrackedEntityInstanceService trackedEntityInstanceService;
+
+    @Autowired
+    private IdentifiableObjectManager manager;
+
+    @Autowired
+    private JdbcMaintenanceStore jdbcMaintenanceStore;
+
+    @Test
+    public void testHardDeleteTei()
+    {
+        OrganisationUnit ou = createOrganisationUnit( 'A' );
+        TrackedEntityAttribute attribute = createTrackedEntityAttribute( 'A' );
+        manager.save( ou );
+        manager.save( attribute );
+
+        TrackedEntityInstance tei = createTrackedEntityInstance( 'A', ou, attribute );
+
+        trackedEntityInstanceService.addTrackedEntityInstance( tei );
+
+        trackedEntityInstanceService.deleteTrackedEntityInstance( tei );
+
+        final AuditQuery query = AuditQuery.builder()
+            .uid( Sets.newHashSet( tei.getUid() ) )
+            .build();
+
+        await().atMost( TIMEOUT, TimeUnit.SECONDS ).until( () -> auditService.countAudits( query ) > 0 );
```

Comment:
I assume this is basically to wait for our embedded artemis to have processed the message? is there a better way here? maybe using mocking? @luciano-fiandesio

Reply:
@mortenoh I suggested @vietnguyen to use this library that is designed exactly for these kind of async tests. Since this is not an unit test, we can't really use mocks.